### PR TITLE
sql: override StmtTimeout to 0 for background jobs

### LINKED ISSUE #126261: backfill: override statement_timeout to 0 in VERIFY INDEX job
Currently, for index validation jobs kicked off in schema changes like CREATE INDEX, we inherit the session variables used by the internal executor from the global default (i.e. whichever value is set by the cluster setting). This means that we can land in situations like the following:

```
root@localhost:29000/movr> SET CLUSTER SETTING sql.defaults.statement_timeout = '5ms';                                             
NOTICE: setting global default sql.defaults.statement_timeout is not recommended
HINT: use the `ALTER ROLE ... SET` syntax to control session variable defaults at a finer-grained level. See: https://www.cockroachlabs.com/docs/v23.2/alter-role.html#set-default-session-variable-values-for-a-role
SET CLUSTER SETTING

Time: 3ms total (execution 2ms / network 0ms)

root@localhost:29000/movr> SET statement_timeout = '0s';                                                                           
SET

Time: 0ms total (execution 0ms / network 0ms)

root@localhost:29000/movr> SHOW statement_timeout;                                                                                 
  statement_timeout
---------------------
  0
(1 row)

Time: 0ms total (execution 0ms / network 0ms)

root@localhost:29000/movr> CREATE INDEX test on public.users (credit_card, id);                                                    
ERROR: job 980684294138527745 could not be loaded: failed to get payload for job 980684294138527745: job-info-get: query execution canceled due to statement timeout
SQLSTATE: 57014
HINT: The job may not have succeeded.
```

Jobs kicked off by schema changes should not need to respect this timeout -- which was the behavior pre-23.2.

We can set a SessionDataOverride that overrides statement_timeout to '0s' for this job.

Jira issue: CRDB-39824

Epic CRDB-40419

### PR DESCRIPTION
**Pull request by annrpom**
For background jobs and schema changes, our Internal
Executor inherits from cluster settings. If a statement
timeout is set cluster-wide, then background jobs will
respect this -- which does not make much sense; thus,
this patch overrides the StmtTimeout to 0s for background
jobs.

Fixes: https://github.com/cockroachdb/cockroach/issues/126261
Release note (bug fix): Background jobs no longer respect
a statement timeout.

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126336)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @mgartner, and @rafiss)

___
*[`pkg/sql/internal.go` line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126336#-O0QJNyA-oS4hxf2o9Of:-O0QJNyA-oS4hxf2o9Og:b-23vu7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/e15721545b8a749f3f59b411aa48d7f166e47a73/pkg/sql/internal.go#L98)):*
> ```Go
> 	sd.SequenceState = sessiondata.NewSequenceState()
> 	sd.Location = time.UTC
> 	sd.StmtTimeout = 0
> ```

Hm, this would disable the statement timeout effectively for all queries issued via the internal executor, not only the background jobs and schema changes. I wonder whether this is desirable or not. As a concrete example, I think this would disable the stmt timeout for queries issued via the Cloud shell - search for `run-query-via-api` for the callsite.

One somewhat similar precedent we have is the desire to disable distsql for all internal queries in the jobs system. This is achieved via `NewInternalDBWithSessionDataOverrides`. Perhaps we want to do something like that instead? I.e. add another override for the stmt timeout in `server_sql.go` and find the relevant InternalDB used by the schema changes?

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126336)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @yuzefovich)

___
*[`pkg/sql/internal.go` line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126336#-O0QJNyA-oS4hxf2o9Of:-O0U93bk19tMp7reDNvg:b-1rl8bt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e15721545b8a749f3f59b411aa48d7f166e47a73/pkg/sql/internal.go#L98)):*
> Hm, this would disable the statement timeout effectively for all queries issued via the internal executor, not only the background jobs and schema changes.

I don't believe this is the case. Annie added a test that shows the session-bound IEs do inherit the value:
```
# Ensure that a session-bound IE still inherits from session vars, if available;
# otherwise, it inherits from the cluster setting.
query T
SELECT crdb_internal.execute_internally('SHOW statement_timeout;', true);
----
39600000
```

as per the comment on this `NewInternalSessionData` function, this is only called for IE queries that are not bound to a session
```
// NewInternalSessionData returns a session data for use in internal queries
// that are not run on behalf of a user session, such as those run during the
// steps of background jobs and schema changes. Each session variable is
// initialized using the correct default value.
```

this is the relevant part of the code: https://github.com/cockroachdb/cockroach/blob/b9680b45d61f497fe996ebefda052d8c4e5b9eb3/pkg/sql/internal.go#L1132-L1137

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126336)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rafiss)

___
*[`pkg/sql/internal.go` line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126336#-O0QJNyA-oS4hxf2o9Of:-O0UUlMRB2RTU2vVPD4E:b-w3xenb) ([raw file](https://github.com/cockroachdb/cockroach/blob/e15721545b8a749f3f59b411aa48d7f166e47a73/pkg/sql/internal.go#L98)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

> Hm, this would disable the statement timeout effectively for all queries issued via the internal executor, not only the background jobs and schema changes.

I don't believe this is the case. Annie added a test that shows the session-bound IEs do inherit the value:
```
# Ensure that a session-bound IE still inherits from session vars, if available;
# otherwise, it inherits from the cluster setting.
query T
SELECT crdb_internal.execute_internally('SHOW statement_timeout;', true);
----
39600000
```

as per the comment on this `NewInternalSessionData` function, this is only called for IE queries that are not bound to a session
```
// NewInternalSessionData returns a session data for use in internal queries
// that are not run on behalf of a user session, such as those run during the
// steps of background jobs and schema changes. Each session variable is
// initialized using the correct default value.
```

this is the relevant part of the code: https://github.com/cockroachdb/cockroach/blob/b9680b45d61f497fe996ebefda052d8c4e5b9eb3/pkg/sql/internal.go#L1132-L1137
</blockquote></details>

I think it depends on how the internal executor is used. `crdb_internal.execute_internally` builtin uses the "session-bound" mode via `planner` methods, but this is not the only way the IE is exposed.

For example, we get 0 timeout in the `run-query-via-api` which doesn't seem desirable. I applied the following diffs:
```
--- a/pkg/server/api_v2_sql_test.go
+++ b/pkg/server/api_v2_sql_test.go
@@ -41,6 +41,10 @@ func TestExecSQL(t *testing.T) {
        ctx := context.Background()
        defer server.Stopper().Stop(ctx)
 
+       sqlDB := server.ApplicationLayer().SQLConn(t)
+       _, err := sqlDB.ExecContext(ctx, "SET CLUSTER SETTING sql.defaults.statement_timeout = '36000000ms';")
+       require.NoError(t, err)
+
        adminClient, err := server.GetAdminHTTPClient()
        require.NoError(t, err)
...
--- a/pkg/sql/internal.go
+++ b/pkg/sql/internal.go
@@ -95,6 +95,7 @@ func NewInternalSessionData(
        sd.SearchPath = sessiondata.DefaultSearchPathForUser(username.NodeUserName())
        sd.SequenceState = sessiondata.NewSequenceState()
        sd.Location = time.UTC
+       sd.StmtTimeout = 0
        return sd
 }
```
and then adjusted `pkg/server/testdata/api_v2_sql` to run `SHOW statement_timeout;` via `sql admin` and `sql non-admin` directives. In both cases we get 0 value.

This is because we access the IE via `sqlServer.internalDB.Txn` call which isn't session bound, so we don't inherit the session data. If we get access to IE via `InternalDB.Executor` call without specifying `WithSessionData` option, I think we also don't inherit it.

In short, IIUC the comment saying that `NewInternalSessionData` is used only for background jobs and schema changes seems aspirational to me - we have places where it ends up being used implicitly because of the code you linked (if `InternalExecutor` doesn't have `sessionDataStack` set, then we always call `NewInternalSessionData`).

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126336)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rafiss)

___
*[`pkg/sql/internal.go` line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126336#-O0QJNyA-oS4hxf2o9Of:-O0UbzBD9p_VTijdGOuF:bx8k5jr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e15721545b8a749f3f59b411aa48d7f166e47a73/pkg/sql/internal.go#L98)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I think it depends on how the internal executor is used. `crdb_internal.execute_internally` builtin uses the "session-bound" mode via `planner` methods, but this is not the only way the IE is exposed.

For example, we get 0 timeout in the `run-query-via-api` which doesn't seem desirable. I applied the following diffs:
```
--- a/pkg/server/api_v2_sql_test.go
+++ b/pkg/server/api_v2_sql_test.go
@@ -41,6 +41,10 @@ func TestExecSQL(t *testing.T) {
        ctx := context.Background()
        defer server.Stopper().Stop(ctx)
 
+       sqlDB := server.ApplicationLayer().SQLConn(t)
+       _, err := sqlDB.ExecContext(ctx, "SET CLUSTER SETTING sql.defaults.statement_timeout = '36000000ms';")
+       require.NoError(t, err)
+
        adminClient, err := server.GetAdminHTTPClient()
        require.NoError(t, err)
...
--- a/pkg/sql/internal.go
+++ b/pkg/sql/internal.go
@@ -95,6 +95,7 @@ func NewInternalSessionData(
        sd.SearchPath = sessiondata.DefaultSearchPathForUser(username.NodeUserName())
        sd.SequenceState = sessiondata.NewSequenceState()
        sd.Location = time.UTC
+       sd.StmtTimeout = 0
        return sd
 }
```
and then adjusted `pkg/server/testdata/api_v2_sql` to run `SHOW statement_timeout;` via `sql admin` and `sql non-admin` directives. In both cases we get 0 value.

This is because we access the IE via `sqlServer.internalDB.Txn` call which isn't session bound, so we don't inherit the session data. If we get access to IE via `InternalDB.Executor` call without specifying `WithSessionData` option, I think we also don't inherit it.

In short, IIUC the comment saying that `NewInternalSessionData` is used only for background jobs and schema changes seems aspirational to me - we have places where it ends up being used implicitly because of the code you linked (if `InternalExecutor` doesn't have `sessionDataStack` set, then we always call `NewInternalSessionData`).
</blockquote></details>

(Should've explicitly mentioned right away: removing `sd.StmtTimeout = 0` diff in the test above makes it so that non-zero value is returned on `SHOW statement_timeout` as expected.)

**Comment by annrpom** [comment]
Coming up with some tests

**Review by rafiss — CHANGES_REQUESTED** [review]
changes look good! i just had some nits

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126336#-:-O0uJO242aECEhs3NBdV:b-vnvlht)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @mgartner, and @yuzefovich)

___
*[`pkg/sql/create_index_test.go` line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126336#-O0uIdsa8s7SJEcsNdRT:-O0uIdsa8s7SJEcsNdRU:by290uf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ace031953ba23e9e8b53b6edd9a198385a9094c7/pkg/sql/create_index_test.go#L11)):*
> ```Go
> // licenses/APL.txt.
> 
> package sql_test
> ```

super nits:
- for bigger integrationy tests like this that don't really test a specific feature, we sometimes put those in pkg/sql/tests. we're pretty unprincipled about this though, but seems like a good idea.
- the file name might be better off describing the bug that we're adding a regression test for. maybe `job_statement_timeout_test.go`.
- same for the test name, maybe `TestBackgroundJobIgnoresStatementTimeout`. (the reason being, we don't really care about `create index` specifically here.)
___
*[`pkg/sql/create_index_test.go` line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126336#-O0uHsvaDLKJdEZSvobw:-O0uHsvb4Z2aWgfFDCDy:b-v3vrjf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ace031953ba23e9e8b53b6edd9a198385a9094c7/pkg/sql/create_index_test.go#L58)):*
> ```Go
> 
> 	var jobCount int
> 	// We don't do it earlier because migration-job-find-already-completed
> ```

nit: is the comment incomplete?
___
*[`pkg/sql/create_index_test.go` line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126336#-O0uJBieC1fS_7bvgdgZ:-O0uJBieC1fS_7bvgdg_:b4x8gpw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ace031953ba23e9e8b53b6edd9a198385a9094c7/pkg/sql/create_index_test.go#L60)):*
> ```Go
> 	// We don't do it earlier because migration-job-find-already-completed
> 	tDB.Exec(t, "SET CLUSTER SETTING sql.defaults.statement_timeout = '3s'")
> 	tDB.Exec(t, `CREATE UNIQUE INDEX bar ON t (b)`)
> ```

nit: this should be a `SucceedsSoon`. otherwise it will fail if the job doesn't complete immediately.
___
*[`pkg/sql/create_index_test.go` line 63 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126336#-O0uJG5t8NrVwPvKn1ej:-O0uJG5t8NrVwPvKn1ek:b-n2kup8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ace031953ba23e9e8b53b6edd9a198385a9094c7/pkg/sql/create_index_test.go#L63)):*
> ```Go
> 	q := `SELECT COUNT(*) 
> FROM crdb_internal.jobs 
> WHERE job_type = 'NEW SCHEMA CHANGE' AND status = 'succeeded'`
> ```

nit: i think the WHERE can be a bit more specific so that it looks for our `CREATE INDEX` job rather than any schema change job

**Review by annrpom — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126336)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, and @yuzefovich)

___
*[`pkg/sql/create_index_test.go` line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126336#-O0uIdsa8s7SJEcsNdRT:-O0uJoBsCvpgv89HsjyR:brafuwn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ace031953ba23e9e8b53b6edd9a198385a9094c7/pkg/sql/create_index_test.go#L11)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nits:
- for bigger integrationy tests like this that don't really test a specific feature, we sometimes put those in pkg/sql/tests. we're pretty unprincipled about this though, but seems like a good idea.
- the file name might be better off describing the bug that we're adding a regression test for. maybe `job_statement_timeout_test.go`.
- same for the test name, maybe `TestBackgroundJobIgnoresStatementTimeout`. (the reason being, we don't really care about `create index` specifically here.)
</blockquote></details>

ahhh this makes sense -- thanks for the context
___
*[`pkg/sql/create_index_test.go` line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126336#-O0uHsvaDLKJdEZSvobw:-O0uKO7o68AXzWqaYq6v:b-grx5w5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ace031953ba23e9e8b53b6edd9a198385a9094c7/pkg/sql/create_index_test.go#L58)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: is the comment incomplete?
</blockquote></details>

lost the comment in a squash ty for catching
___
*[`pkg/sql/create_index_test.go` line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126336#-O0uJBieC1fS_7bvgdgZ:-O0uL-V6DkVRezJ8n20X:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ace031953ba23e9e8b53b6edd9a198385a9094c7/pkg/sql/create_index_test.go#L60)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: this should be a `SucceedsSoon`. otherwise it will fail if the job doesn't complete immediately.
</blockquote></details>

done
___
*[`pkg/sql/create_index_test.go` line 63 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126336#-O0uJG5t8NrVwPvKn1ej:-O0uL9Xg6nqvV4JP6MnE:bbwjjqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ace031953ba23e9e8b53b6edd9a198385a9094c7/pkg/sql/create_index_test.go#L63)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: i think the WHERE can be a bit more specific so that it looks for our `CREATE INDEX` job rather than any schema change job
</blockquote></details>

sure done -- better to be more explicit about what we're searching for here. qq: do we do these test in a legacy schema changer only env as well? stressing this always gave me DSC so just double-checking

**Review by rafiss — APPROVED** [review]
that's a good point, we'd need to add an explicit test for legacy schema changer. take a look at the testutils.RunTrueAndFalse helper.

feel free to merge once that test is added!

**Comment by annrpom** [comment]
tftrs!

bors r+
